### PR TITLE
ENH/MAINT: HiGHS upstream enhancements and bug fixes

### PR DESCRIPTION
**Pull request by mckib2**
#### What does this implement/fix?

- All C/C++ source/header changes due to upstream HiGHS development
- `HighsModelStatus` enum definitions moved to different header
- introduces `start_crossover_tolerance` for `_highs_wrapper.pyx`
- renames default edge weight strategy `SIMPLEX_DUAL_EDGE_WEIGHT_STRATEGY_STEEPEST_EDGE_TO_DEVEX_SWITCH` to `SIMPLEX_DUAL_EDGE_WEIGHT_STRATEGY_STEEPEST_EDGE_UNIT_INITIAL`
- Additions to PYD files to keep up with HiGHS API
- Handle new HiGHS primal and primal/dual in-feasibility statuses
- update `_highs_wrapper.pyx` and `_highs_constants.pyx` to reflect the above
- fix benchmarks to use `method="highs-ds"` instead of `method="highs-simplex"`

#### Additional information

- no public `linprog` API changes
- failing tests due to presolve, see upstream https://github.com/ERGO-Code/HiGHS/issues/425 _EDIT_: fixed

### CODE REVIEW

**Comment by mckib2** [comment]
@mdhaber Opening PR now even though merging is blocked due to failed tests to solicit any other feedback while waiting for upstream fixes

**Comment by mckib2** [comment]
All tests are now passing, @mdhaber can you take a look?

**Comment by mckib2** [comment]
Merged in current SciPy master, still on HiGHS circa November of 2020 (which is fine, I'd like to find a cadence to do more regular updates)

**Comment by mckib2** [comment]
@rgommers @mdhaber Do you think it would be useful to take the approach done with Boost in gh-13328 and provide the HiGHS as a zip distribution?  That would prevent the diffs from being clogged up with vendor-controlled source changes that we're not interested in reviewing and let reviewers focus on SciPy-specific changes

**Comment by rgommers** [comment]
Putting source code in a zip file seems like an unhealthy plan. That just makes it harder to debug and work with within the repo I'd think. We don't do this for any other vendored libraries. PR review in cases like these can be aided my keeping separate commits for adding vendored sources, for patching those sources if needed (ideally making it easy to re-apply SciPy-specific patches when the vendored code gets updated), and commits for bindings and Python code.

File: benchmarks/benchmarks/optimize_linprog.py [hunk 1]
```diff
@@ -37,7 +37,7 @@
            ("interior-point", {"sparse": False}),
            ("revised simplex", {}),
            ("highs-ipm", {}),
-           ("highs-simplex", {})]
+           ("highs-ds", {})]
```
Comment (mdhaber) [inline benchmarks/benchmarks/optimize_linprog.py:40]: Thanks; I noticed this and meant to do that.

File: scipy/optimize/_linprog_highs.py [hunk 2]
```diff
@@ -211,7 +212,7 @@ def _linprog_highs(lp, solver, time_limit=None, presolve=True,
         'simplex_dual_edge_weight_strategy',
         choices={'dantzig': HIGHS_SIMPLEX_DUAL_EDGE_WEIGHT_STRATEGY_DANTZIG,
                  'devex': HIGHS_SIMPLEX_DUAL_EDGE_WEIGHT_STRATEGY_DEVEX,
-                 'steepest-devex': HIGHS_SIMPLEX_DUAL_EDGE_WEIGHT_STEEP2DVX,
+                 'steepest-devex': HIGHS_SIMPLEX_DUAL_EDGE_WEIGHT_STRATEGY_CHOOSE,
```
Comment (mdhaber) [inline scipy/optimize/_linprog_highs.py:215]: Did they remove steep2dvx? Is this equivalent?

File: scipy/optimize/_linprog_highs.py [hunk 3]
```diff
@@ -245,6 +242,10 @@ def _linprog_highs(lp, solver, time_limit=None, presolve=True,
             4,
             'HiGHS Status Code 6: HighsModelStatusPOSTSOLVE_ERROR',
         ),
+        MODEL_STATUS_MODEL_EMPTY: (
+            4,
+            'HiGHS Status Code 3: HighsModelStatusMODEL_EMPTY',
+        ),
```
Comment (mdhaber) [inline scipy/optimize/_linprog_highs.py:248]: Was moving this intentional? Any functional change?

File: scipy/optimize/_linprog_highs.py (hunk 2 above)
Comment (mckib2) [reply scipy/optimize/_linprog_highs.py:215]: This is equivalent to what was there before, this was a renaming of the enumeration.  This is verified by checking the [source](https://github.com/ERGO-Code/HiGHS/blob/25c112b29520b4ecb1224e33c1f42471dbe51cbe/src/simplex/HDual.cpp#L1919) where `allow_dual_steepest_edge_to_devex_switch` is set to `true`

File: scipy/optimize/_linprog_highs.py (hunk 3 above)
Comment (mckib2) [reply scipy/optimize/_linprog_highs.py:248]: Yes, it was intentional to match the reordering of the enumeration.  There is no functional change, just makes it easier to compare when it's in the same order

**Comment by mdhaber** [comment]
OK. I'm just going to run the benchmarks locally. Maybe compare with master. If that looks good I'll merge.

**Comment by mdhaber** [comment]
I'll go ahead and merge this to keep HiGHS up to date. There doesn't seem to be much change in performance. These are the only things about HiGHS that changed.
```
       before           after         ratio
     [09ef43de]       [7bb26861]
     <master>         <highs-updtream-updates-2020-11-28>
+               2                4     2.00  optimize_linprog.Netlib_infeasible.track_netlib_infeasible(('highs-ds', {}), 'qual')
+          failed                4     2.00  optimize_linprog.Netlib_infeasible.track_netlib_infeasible(('highs-ds', {}), 'refinery')
+          failed                4     2.00  optimize_linprog.Netlib_infeasible.track_netlib_infeasible(('highs-ds', {}), 'vol1')
-     1.08±0.05ms         955±10μs     0.89  optimize_linprog.Netlib_infeasible.time_netlib_infeasible(('highs-ds', {}), 'ex73a')
-      11.5±0.2ms       9.82±0.2ms     0.85  optimize_linprog.Netlib_infeasible.time_netlib_infeasible(('highs-ipm', {}), 'qual')
```
The first of these is a regression (problem is infeasible - 2 - but is now running into some sort of problem - 4). The other two are improvements. We should investigate and possibly report the `'qual'` regression upstream.

[Here](https://www.dropbox.com/s/7czkgsmldgtvce0/13197vsMaster.txt?dl=0) are the full results.

Two other thoughts that should probably be copied somewhere else:
- I think it's time to take `"interior-point"` and `"revised simplex"` out of the benchmark suite entirely. I don't think we'll be investing time into improving them, so they just slow down the benchmark suite.
- `'highs-ds'` fails ETAMACRO when tracking accuracy but not when timing it. If it solves this problem when executed in a normal Python session, I wouldn't worry about it; this wouldn't be the first time benchmarks have reported results that are totally mystifying. Perhaps there is an error log I've not learned to read...

**Pull request merged**